sysboundary/ionosphere.cpp

```diff
@@ -851,13 +879,288 @@ namespace SBC {
+            didOnePopulationAtLeast = true;
+            
+            std::vector<Real> sumWeights(productionNumProtonEnergies, 0);
+            
+            # pragma omp parallel
+            {
+               std::vector<Real> thread_lossCone_sum(productionNumProtonEnergies,0.0);
+               std::vector<Real> thread_count(productionNumProtonEnergies,0.0);
+               
+               const Real* parameters  = cell->get_block_parameters(popID);
+               const Realf* block_data = cell->get_data(popID);
+               
+               # pragma omp for
+               for (vmesh::LocalID vn=0; vn<cell->get_number_of_velocity_blocks(popID); vn++) {
+                  for (uint k = 0; k < WID; ++k) for (uint j = 0; j < WID; ++j) for (uint i = 0; i < WID; ++i) {
+                     const Real VX 
+                     =           parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::VXCRD] 
+                     + (i + 0.5)*parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVX];
+                     const Real VY 
+                     =           parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::VYCRD] 
+                     + (j + 0.5)*parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVY];
+                     const Real VZ 
+                     =           parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::VZCRD] 
+                     + (k + 0.5)*parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVZ];
+                     
+                     const Real DV3 
+                     = parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVX]
+                     * parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVY] 
+                     * parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVZ];
+                     
+                     const Real normV = sqrt(VX*VX + VY*VY + VZ*VZ);
+                     const Real VdotB_norm = (B[0]*VX + B[1]*VY + B[2]*VZ)/normV;
+                     Real countAndGate = floor(VdotB_norm/cosAngle);  // gate function: 0 outside loss cone, 1 inside
+                     countAndGate = max(0.,countAndGate);
+
+                     const Real energy = 0.5 * mass * normV*normV / (1000 * physicalconstants::CHARGE); // in keV, as the limits of the bins
+                     
+                     // Find the correct energy bin number to update
+                     int binNumber = round((log(energy) - log(productionMinProtonEnergy)) / log(productionMaxProtonEnergy/productionMinProtonEnergy) * (productionNumProtonEnergies-1));
+                     binNumber = max(binNumber, 0); // anything < emin goes to the lowest channel
+                     binNumber = min(binNumber, productionNumProtonEnergies-1); // anything > emax goes to the highest channel
+                     
+                     // TODO check the bin number or -1
```

Comment:
So I think that on 1022 it should not have a minus 1 (that is copied from Maxime's code where the indexing, I believe, is handled slightly differently – possibly incorrectly there imho).